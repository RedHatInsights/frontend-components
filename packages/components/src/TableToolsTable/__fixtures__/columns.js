import { severityLevels } from './items';

export default [
  {
    title: 'ID',
    sortByProperty: 'id',
    exportKey: 'id',
  },
  {
    title: 'Name',
    sortByProperty: 'name',
    exportKey: 'name',
    renderFunc: (_a, _b, item) => item.name,
  },
  {
    title: 'Another Name column',
    sortByProperty: 'name',
    exportKey: 'name',
    renderExport: (name) => `${name} via export render`,
  },
  {
    title: 'Description',
    sortByProperty: 'description',
    renderExport: (item) => `${item.name} description rendered for export`,
  },
  {
    title: 'Severity sorted by function',
    sortByFunction: (item) => item.name,
    exportKey: 'severity',
    renderFunc: (_a, _b, item) => item.severity,
  },
  {
    title: 'Severity by Array',
    sortByProperty: 'severity',
    sortByArray: severityLevels,
    exportKey: 'severity',
    renderFunc: (_a, _b, item) => item.severity,
  },
];
