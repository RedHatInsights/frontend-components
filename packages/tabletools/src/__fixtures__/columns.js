import { severityLevels } from './items';

export default [
  {
    title: 'ID',
    sortByProperty: 'id',
    exportKey: 'id',
    key: 'id',
    managable: false,
  },
  {
    title: 'Name',
    key: 'id',
    sortByProperty: 'name',
    exportKey: 'name',
    renderFunc: (_a, _b, item) => item.name,
  },
  {
    title: 'Another Name column',
    key: 'another_name_column',
    sortByProperty: 'name',
    exportKey: 'name',
    renderExport: (name) => `${name} via export render`,
  },
  {
    title: 'Description',
    key: 'desc',
    sortByProperty: 'description',
    renderExport: (item) => `${item.name} description rendered for export`,
  },
  {
    title: 'Severity sorted by function',
    sortByFunction: (item) => item.name,
    exportKey: 'severity',
    renderFunc: (_a, _b, item) => item.severity,
  },
  {
    title: 'Severity by Array',
    sortByProperty: 'severity',
    sortByArray: severityLevels,
    exportKey: 'severity',
    renderFunc: (_a, _b, item) => item.severity,
  },
];
