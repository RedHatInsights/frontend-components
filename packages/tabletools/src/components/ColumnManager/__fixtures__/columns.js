export default [
  {
    title: 'ID',
    key: 'id',
  },
  {
    title: 'Another col',
    key: 'name',
    isRequired: true,
  },
  {
    title: 'Name',
    key: 'name1',
  },
  {
    title: 'Another Name column without key',
  },
];
