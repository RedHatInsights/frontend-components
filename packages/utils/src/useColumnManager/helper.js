import get from 'lodash/get';

export const filterColumnsBySelected = (columns = [], selected = false, selectProp = '') =>
  columns.filter((column) => selected.includes(get(column, selectProp)));

const managableColumns = (columns = []) =>
  columns.map((column) => (column.managable === undefined ? { ...column, managable: true } : column)).filter((column) => column.managable === true);

export const filterManageableColumns = (columns = [], prop = '') => managableColumns(columns.filter((column) => column.?[prop] !== undefined));
