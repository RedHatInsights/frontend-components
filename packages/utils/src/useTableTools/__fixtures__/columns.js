import { severityLevels } from './items';

export default [
    {
        title: 'Name',
        key: 'name',
        sortByProperty: 'name',
        renderFunc: (item) => (
            item.name
        )
    }, {
        title: 'Description',
        key: 'description'
    }, {
        title: 'Severity sorted by function',
        sortByFunction: (item) => (severityLevels.indexOf(item.severity)),
        renderFunc: (item) => (
            item.severity
        )
    }, {
        title: 'Severity by Array',
        sortByProperty: 'severity',
        sortByArray: severityLevels,
        renderFunc: (item) => (
            item.severity
        )
    }
];
