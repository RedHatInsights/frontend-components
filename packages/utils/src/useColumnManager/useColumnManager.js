import React, { useState } from 'react';
import get from 'lodash/get';
import ColumnManager from '@redhat-cloud-services/frontend-components/ColumnManager';
import { filterColumnsBySelected, filterManageableColumns } from './helper';

const useColumnManager = (columns = [], options = {}) => {
  const { columnManagerSelectProp: selectProp = 'key', manageColumns: enableColumnManager, manageColumnLabel = 'Manage columns' } = options;
  const managableColumns = filterManageableColumns(columns, selectProp);
  const [selectedColumns, setSelectedColumns] = useState(managableColumns.map((column) => get(column, selectProp)));
  const [isManagerOpen, setIsManagerOpen] = useState(false);

  const onClick = () => {
    setIsManagerOpen(true);
  };

  const onSave = (newSelectedColumns) => {
    setSelectedColumns(newSelectedColumns);
    setIsManagerOpen(false);
  };

  return enableColumnManager && managableColumns.length !== 0
    ? {
        columns: filterColumnsBySelected(managableColumns, selectedColumns, selectProp),
        columnManagerAction: {
          label: manageColumnLabel,
          onClick,
        },
        // eslint-disable-next-line react/display-name
        ColumnManager: () => (
          <ColumnManager
            columns={managableColumns}
            isOpen={isManagerOpen}
            onClose={() => setIsManagerOpen(false)}
            selectedColumns={selectedColumns}
            onSave={onSave}
            selectProp={selectProp}
          />
        ),
      }
    : { columns };
};

export default useColumnManager;
