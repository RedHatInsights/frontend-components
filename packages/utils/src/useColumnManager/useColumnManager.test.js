import { renderHook } from '@testing-library/react-hooks';
import columns from '../__fixtures__/columns';
import useColumnManager from './useColumnManager';

describe('useColumnManager', () => {
  const defaultArguments = [columns, { manageColumns: true }];
  it('returns just columns if not enabled', () => {
    const { result } = renderHook(() => useColumnManager(columns));
    expect(result).toMatchSnapshot();
  });

  it('returns just columns if not enabled', () => {
    const { result } = renderHook(() => useColumnManager(...defaultArguments));
    expect(result).toMatchSnapshot();
  });
});
