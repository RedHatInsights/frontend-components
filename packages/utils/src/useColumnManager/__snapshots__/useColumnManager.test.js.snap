// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useColumnManager returns just columns if not enabled 1`] = `
Object {
  "all": Array [
    Object {
      "columns": Array [
        Object {
          "exportKey": "id",
          "key": "id",
          "managable": false,
          "sortByProperty": "id",
          "title": "ID",
        },
        Object {
          "exportKey": "name",
          "key": "id",
          "renderFunc": [Function],
          "sortByProperty": "name",
          "title": "Name",
        },
        Object {
          "exportKey": "name",
          "key": "another_name_column",
          "renderExport": [Function],
          "sortByProperty": "name",
          "title": "Another Name column",
        },
        Object {
          "key": "desc",
          "renderExport": [Function],
          "sortByProperty": "description",
          "title": "Description",
        },
        Object {
          "exportKey": "severity",
          "renderFunc": [Function],
          "sortByFunction": [Function],
          "title": "Severity sorted by function",
        },
        Object {
          "exportKey": "severity",
          "renderFunc": [Function],
          "sortByArray": Array [
            "high",
            "medium",
            "low",
            "unknown",
          ],
          "sortByProperty": "severity",
          "title": "Severity by Array",
        },
      ],
    },
  ],
  "current": Object {
    "columns": Array [
      Object {
        "exportKey": "id",
        "key": "id",
        "managable": false,
        "sortByProperty": "id",
        "title": "ID",
      },
      Object {
        "exportKey": "name",
        "key": "id",
        "renderFunc": [Function],
        "sortByProperty": "name",
        "title": "Name",
      },
      Object {
        "exportKey": "name",
        "key": "another_name_column",
        "renderExport": [Function],
        "sortByProperty": "name",
        "title": "Another Name column",
      },
      Object {
        "key": "desc",
        "renderExport": [Function],
        "sortByProperty": "description",
        "title": "Description",
      },
      Object {
        "exportKey": "severity",
        "renderFunc": [Function],
        "sortByFunction": [Function],
        "title": "Severity sorted by function",
      },
      Object {
        "exportKey": "severity",
        "renderFunc": [Function],
        "sortByArray": Array [
          "high",
          "medium",
          "low",
          "unknown",
        ],
        "sortByProperty": "severity",
        "title": "Severity by Array",
      },
    ],
  },
  "error": undefined,
}
`;

exports[`useColumnManager returns just columns if not enabled 2`] = `
Object {
  "all": Array [
    Object {
      "ColumnManager": [Function],
      "columnManagerAction": Object {
        "label": "Manage columns",
        "onClick": [Function],
      },
      "columns": Array [
        Object {
          "exportKey": "name",
          "key": "id",
          "managable": true,
          "renderFunc": [Function],
          "sortByProperty": "name",
          "title": "Name",
        },
        Object {
          "exportKey": "name",
          "key": "another_name_column",
          "managable": true,
          "renderExport": [Function],
          "sortByProperty": "name",
          "title": "Another Name column",
        },
        Object {
          "key": "desc",
          "managable": true,
          "renderExport": [Function],
          "sortByProperty": "description",
          "title": "Description",
        },
      ],
    },
  ],
  "current": Object {
    "ColumnManager": [Function],
    "columnManagerAction": Object {
      "label": "Manage columns",
      "onClick": [Function],
    },
    "columns": Array [
      Object {
        "exportKey": "name",
        "key": "id",
        "managable": true,
        "renderFunc": [Function],
        "sortByProperty": "name",
        "title": "Name",
      },
      Object {
        "exportKey": "name",
        "key": "another_name_column",
        "managable": true,
        "renderExport": [Function],
        "sortByProperty": "name",
        "title": "Another Name column",
      },
      Object {
        "key": "desc",
        "managable": true,
        "renderExport": [Function],
        "sortByProperty": "description",
        "title": "Description",
      },
    ],
  },
  "error": undefined,
}
`;
